NorthstarDLL/httprequesthandler.cpp

```diff
@@ -0,0 +1,602 @@
+				}
+			}
+
+			curl_easy_cleanup(curl);
+			curl_slist_free_all(headers);
+			curl_slist_free_all(host);
+		});
+
+	requestThread.detach();
+	return handle;
+}
+
+template <ScriptContext context> void HttpRequestHandler::RegisterSQFuncs()
+{
+	g_pSquirrel<context>->AddFuncRegistration(
+		"int",
+		"NS_InternalMakeHttpRequest",
+		"int method, string baseUrl, table<string, array<string> > headers, table<string, array<string> > queryParams, string contentType, "
+		"string body, "
+		"int timeout, string userAgent",
+		"[Internal use only] Passes the HttpRequest struct fields to be reconstructed in native and used for an http request",
+		SQ_InternalMakeHttpRequest<context>);
+
+	g_pSquirrel<context>->AddFuncRegistration(
+		"bool",
+		"NSIsHttpEnabled",
+		"",
+		"Whether or not HTTP requests are enabled. You can opt-out by starting the game with -disablehttprequests.",
+		SQ_IsHttpEnabled<context>);
+
+	g_pSquirrel<context>->AddFuncRegistration(
+		"bool",
+		"NSIsLocalHttpAllowed",
+		"",
+		"Whether or not HTTP requests can be made to a private network address. You can enable this by starting the game with "
+		"-allowlocalhttp.",
+		SQ_IsLocalHttpAllowed<context>);
+}
+
+// int NS_InternalMakeHttpRequest(int method, string baseUrl, table<string, string> headers, table<string, string> queryParams,
+//	string contentType, string body, int timeout, string userAgent)
+template <ScriptContext context> SQRESULT SQ_InternalMakeHttpRequest(HSquirrelVM* sqvm)
+{
+	if (!g_httpRequestHandler || !g_httpRequestHandler->IsRunning())
+	{
+		spdlog::warn("NS_InternalMakeHttpRequest called while the http request handler isn't running.");
+		g_pSquirrel<context>->pushinteger(sqvm, -1);
+		return SQRESULT_NOTNULL;
+	}
+
+	if (IsHttpDisabled())
+	{
+		spdlog::warn("NS_InternalMakeHttpRequest called while the game is running with -disablehttprequests."
+					 " Please check if requests are allowed using NSIsHttpEnabled() first.");
+		g_pSquirrel<context>->pushinteger(sqvm, -1);
+		return SQRESULT_NOTNULL;
+	}
+
+	HttpRequest request;
+	request.method = static_cast<HttpRequestMethod::Type>(g_pSquirrel<context>->getinteger(sqvm, 1));
+	request.baseUrl = g_pSquirrel<context>->getstring(sqvm, 2);
+
+	// Read the tables for headers and query parameters.
+	SQTable* headerTable = sqvm->_stackOfCurrentFunction[3]._VAL.asTable;
+	for (int idx = 0; idx < headerTable->_numOfNodes; ++idx)
+	{
+		tableNode* node = &headerTable->_nodes[idx];
+
+		// TODO: Figure out why the first three nodes of this table are OT_NULL with a single value.
+		// Then re-enable this if possible.
+		// if (node->key._Type != OT_STRING || node->val._Type != OT_STRING)
+		//{
+		//	g_pSquirrel<context>->raiseerror(
+		//		sqvm, fmt::format("Invalid header type or value (expected string, got {} = {})", SQTypeNameFromID(node->key._Type),
+		// SQTypeNameFromID(node->val._Type)).c_str()); 	return SQRESULT_ERROR;
+		//}
```

Comment:
Is this important?

Reply:
Realistically no since you're forced to provide a `table<string, array<string> >` anyway.

Reply:
I see, these could probably be removed then?

Reply:
Will do.